### PR TITLE
Add protobuf stream data decoder

### PR DESCRIPTION
**Pull request by aman1309**

### CODE REVIEW

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+ * {@link com.linkedin.entitystream.EntityStream} of ByteString.
+ * The implementation is backed by a non blocking {@link ProtobufStreamDataParser}. Because the raw bytes are
+ * pushed to the decoder, it keeps the partially built data structure in a stack.
+ *
+ * @author amgupta1
+ */
+class ProtobufDataDecoder<T extends DataComplex> extends AbstractDataDecoder<T>
+{
+
+  private SymbolTable _symbolTable;
```
Comment (karthikrg) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:52]: final?

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+
+  @Override
+  protected DataComplex createDataList(NonBlockingDataParser parser)
+  {
+    return new DataList(parser.getComplexObjSize());
+  }
+
+  class ProtobufStreamDataParser implements NonBlockingDataParser
+  {
+    private SymbolTable _symbolTable;
```
Comment (karthikrg) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:80]: final

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+    private Deque<Integer> _complexObjTokenSizeStack = new ArrayDeque<>();
+    private int _currComplexObjTokenSize = -1;
+
+    private byte[] _input;  //holds feed input bytes
+    private int _limit;
+    private int _pos;
+
+    private boolean _eofInput;  //no more inputs can be feed if this is set to true
+
+    private TextBuffer _textBuffer;  //buffer to hold parsed string characters.
```
Comment (karthikrg) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:91]: final

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+      {
+        throw new DataDecodingException("Error decoding string reference");
+      }
+      return STRING;
+    }
+
+    private Token finishToken(Token token)
+    {
+      _currentToken = token;
+      if (_currentToken == START_OBJECT)
```
Comment (karthikrg) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:275]: switch instead of if else here?

File: data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  private static DataList decodeList(byte[] bytes) throws Exception
+  {
+    ProtobufDataDecoder<DataList> decoder =
+        new ProtobufDataDecoder<>(null, AbstractDataDecoder.START_ARRAY_TOKEN);
+    return decode(bytes, decoder);
+  }
+
+  private static <T extends DataComplex> T decode(byte[] bytes, ProtobufDataDecoder<T> decoder) throws Exception
+  {
+    Writer<ByteString> writer = new ChunkedByteStringWriter(bytes, 3);
```
Comment (karthikrg) [inline data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java:178]: Does the test pass if you set this to 1? That will be the ultimate litmus test since you would have accounted for all edge cases.

File: data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      throws Exception {
+    ProtobufDataCodec codec = new ProtobufDataCodec(
+        new ProtobufCodecOptions.Builder().setEnableFixedLengthFloatDoubles(enableFixedLengthFloatDoubles)
+            .setEnableASCIIOnlyStrings(false).build());
+    byte[] bytes = TestUtil.dataComplexToBytes(codec, dataComplex);
+    DataComplex decodedDataComplex = decode(bytes);
+    assertEquals(TestUtil.dataComplexToBytes(codec, decodedDataComplex), bytes);
+    codec = new ProtobufDataCodec(
+        new ProtobufCodecOptions.Builder().setEnableFixedLengthFloatDoubles(enableFixedLengthFloatDoubles)
+            .setEnableASCIIOnlyStrings(true).build());
```
Comment (karthikrg) [inline data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java:50]: We should take setEnableASCIIOnlyStrings as a dataprovider param

Comment (aman1309) [reply data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java:50]: actually will remove test cases for setEnableASCIIOnlyStrings=false as this is deprecated, any both ascii and utf-8 string can be tested without this.

File: data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java (hunk 5 above)
Comment (aman1309) [reply data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java:178]: will add test for different buffer size

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+    private int _currComplexObjTokenSize = -1;
+
+    private byte[] _input;  //holds feed input bytes
+    private int _limit;
+    private int _pos;
+
+    private boolean _eofInput;  //no more inputs can be feed if this is set to true
+
+    private final TextBuffer _textBuffer;  //buffer to hold parsed string characters.
+    private int _bufferPos = -1;  //signify no. of chars in text buffers as buffer is reused to avoid thrashing
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:92]: rename this to _textBufferPos

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 8]
```diff
@@ -0,0 +1,891 @@
...
+        throw new DataDecodingException("Unexpected call: Raw bytes value is not available");
+      }
+      return Double.longBitsToDouble(_longValue);
+    }
+
+    /*
+    * Non blocking ProtoReader Implementation
+    */
+
+    private Token readASCIIString() throws IOException {
```
Comment (karthikbalasub) [inline data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:368]: Add doc here, explain the different scenarios:
1. new string, fully readable
2. new string, partially readable
3. reading remaining part of a string.
and include details of the different variables used for each scenario. For example _intValue represents the remaining chars to read for scenario 3.

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 9]
```diff
@@ -0,0 +1,891 @@
...
+      // can simplify some of the checks by short-circuiting right away
+      if (_pos >= _limit) {
+        Token endComplexObjToken = readEndComplexObj();
+        if (endComplexObjToken != NOT_AVAILABLE)
+        {
+          return finishToken(endComplexObjToken);
+        }
+        if (_eofInput) {
+          return EOF_INPUT;
+        }
```
[the remark this answers is not in the document]
Comment (aman1309) [reply data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:154]: after eof there is still a possibility of END object tokens since protobuf doesn't encode end ordinals/tokens

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 10]
```diff
@@ -0,0 +1,891 @@
...
+      if (_currentToken != NOT_AVAILABLE)
+      {
+        _currentToken = readEndComplexObj();
+        if (_currentToken != NOT_AVAILABLE)
+        {
+          return finishToken(_currentToken);
+        }
+        _currentOrdinal = _input[_pos++];
+        //release bytes array if previous token was Token#RAW_BYTES
+        _bytesValue = null;
```
[the remark this answers is not in the document]
Comment (aman1309) [reply data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:166]: I was using _bytesValue as buffer too, to avoid references long byte arrays but now on 2nd thoughts it won't matter much.
Added a byteBuffer separately which can be reset similar to other read pending objects.

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java [hunk 11]
```diff
@@ -0,0 +1,891 @@
...
+    {
+      _currentToken = token;
+      switch (_currentToken)
+      {
+        case START_OBJECT:
+          if (_currComplexObjTokenSize > 0)
+          {
+            _complexObjTokenSizeStack.push(_currComplexObjTokenSize);
+          }
+          _currComplexObjTokenSize = _intValue << 1;
```
[the remark this answers is not in the document]
Comment (aman1309) [reply data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:282]: added comment on stack variable

File: data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java (hunk 8 above)
Comment (aman1309) [reply data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java:368]: added comments for 3 cases.
commented on _intValue variable, also using local variable as remainingSize to signify the same.
